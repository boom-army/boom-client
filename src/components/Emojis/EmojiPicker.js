import { useEffect, useState, useContext } from "react";
import { SmileIcon } from "../Icons";
import { Picker } from "emoji-mart";
import { ThemeContext } from "../../contexts/theme";
import "emoji-mart/css/emoji-mart.css";
import { styled } from "@mui/material/styles";

const PickerWrapper = styled("span")((props) => ({
  ".emoji-mart": {
    position: "absolute",
    zIndex: "1",
  },
  ".emoji-mart-bar.emoji-mart-bar,.emoji-mart-scroll.emoji-mart-scroll": {
    marginRight: "0",
  },
  ".emoji-mart-bar svg,.emoji-mart-bar svg path": {
    fill: props.theme.accentColor,
  },
  ".emoji-mart .emoji-mart-anchor-bar": {
    backgroundColor: `${props.theme.accentColor} !important`,
  },
  ".emoji-mart-preview": {
    display: "none",
  },
  ".emoji-mart-dark": {
    borderColor: props.theme.secondaryColor,
    backgroundColor: props.theme.background,
  },
  ".emoji-mart-category": {
    marginTop: "0.75rem",
  },
  ".emoji-mart-dark .emoji-mart-category-label span": {
    backgroundColor: props.theme.background,
    color: "#fff",
  },
  ".emoji-mart-scroll::WebkitScrollbar": {
    width: "0.25rem",
  },
  ".emoji-mart-scroll::WebkitScrollbarTrack": {
    background: props.theme.background,
  },
  ".emoji-mart-scroll::WebkitScrollbarThumb": {
    background: props.theme.accentColor,
  },
  ".emoji-pick": {
    cursor: "pointer",
  },
  ".emoji-pick:hover svg path": {
    fill: props.theme.accentColor,
  },
  "@media screen and (max-width: 430px)": {
    ".emoji-mart": {
      position: "fixed",
      bottom: "5em",
    },
  },
}));

export const EmojiPicker = ({
  emojiHandler,
  customIcon,
  dismissOnClick,
  props,
}) => {
  const [picker, togglePicker] = useState(false);
  const { theme } = useContext(ThemeContext);

  const handleDocumentClick = (event) => {
    let isEmojiClassFound = false;

    event?.path?.forEach((elem) => {
      if (elem && elem.classList) {
        const data = elem.classList.value;
        if (data.includes("emoji")) {
          isEmojiClassFound = true;
        }
      }
    });
    if (isEmojiClassFound === false && event.target.id !== "emojis-btn")
      togglePicker(false);
  };

  useEffect(() => {
    document.addEventListener("click", handleDocumentClick, false);
    return () => {
      document.removeEventListener("click", handleDocumentClick, false);
    };
  });

  const handleEmojiSelect = (pickedEmoji) => {
    emojiHandler(pickedEmoji);
    if (dismissOnClick) togglePicker(!picker);
  };

  return (
    <PickerWrapper>
      <span className="emoji-pick" onClick={() => togglePicker(!picker)}>
        {customIcon ?? <SmileIcon />}
      </span>

      {picker && (
        <Picker
          button={true}
          sheetSize={64}
          theme={theme.background === "#15202b" ? "dark" : "light"}
          onSelect={handleEmojiSelect}
        />
      )}
    </PickerWrapper>
  );
};
