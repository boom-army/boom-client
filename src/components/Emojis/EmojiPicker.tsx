import { useState, useContext } from "react";
import InsertEmoticonIcon from "@mui/icons-material/InsertEmoticon";
import { Picker } from "emoji-mart";
import { useTheme } from '@mui/material/styles';
import "emoji-mart/css/emoji-mart.css";
import { styled } from "@mui/material/styles";
import { Box, Modal } from "@mui/material";

const PickerWrapper = styled(Box)((props: any) => ({
  position: "absolute" as "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  bgcolor: "background.paper",
  ".emoji-mart-bar svg,.emoji-mart-bar svg path": {
    fill: props.theme.palette.primary.main,
  },
  ".emoji-mart .emoji-mart-anchor-bar": {
    backgroundColor: `${props.theme.palette.primary.main} !important`,
  },
  ".emoji-mart-preview": {
    display: "none",
  },
  ".emoji-mart-dark": {
    borderColor: props.theme.palette.secondary.main,
    backgroundColor: props.theme.palette.background.default,
  },
  ".emoji-mart-dark .emoji-mart-category-label span": {
    backgroundColor: props.theme.palette.background.default,
    color: "#fff",
  },
  ".emoji-mart-scroll::WebkitScrollbar": {
    width: "0.25rem",
  },
  ".emoji-mart-scroll::WebkitScrollbarTrack": {
    background: props.theme.palette.background.default,
  },
  ".emoji-mart-scroll::WebkitScrollbarThumb": {
    background: props.theme.palette.primary.main,
  },
}));

const PickerIcon = styled("span")((props: any) => ({
  cursor: "pointer",
  "& svg path": {
    fill: props.theme.palette.secondary.main,
  },
  "&:hover svg path": {
    fill: props.theme.palette.primary.main,
  },
}));

interface EmojiPickerProps {
  emojiHandler?: any;
  customIcon?: any;
  dismissOnClick?: any;
  props?: any;
}

export const EmojiPicker = ({
  emojiHandler,
  customIcon,
  dismissOnClick,
}: EmojiPickerProps) => {
  const [picker, togglePicker] = useState(false);
  const theme = useTheme();

  const handleEmojiSelect = (pickedEmoji: any) => {
    emojiHandler(pickedEmoji);
    if (dismissOnClick) togglePicker(false);
  };

  return (
    <Box>
      <PickerIcon onClick={() => togglePicker(true)}>
        {customIcon ?? <InsertEmoticonIcon />}
      </PickerIcon>
      <Modal
        open={picker}
        onClose={() => togglePicker(false)}
        aria-labelledby="Emoji picker"
        aria-describedby="Pick an emoji for tweet"
      >
        {/* here change `button` to `useButton` bcz we need to add button props in node modules picker file*/}
        <PickerWrapper>
          {picker && (
            // @ts-ignore
            <Picker
              useButton={true}
              sheetSize={64}
              theme={theme.palette.background.default === "#15202b" ? "dark" : "light"}
              onSelect={handleEmojiSelect}
            />
          )}
        </PickerWrapper>
      </Modal>
    </Box>
  );
};
