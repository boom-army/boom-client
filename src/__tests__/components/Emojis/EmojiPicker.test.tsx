import React from "react";
import { EmojiPicker } from "../../../components/Emojis/EmojiPicker";
import { act, fireEvent, render} from "@testing-library/react";

describe("<EmojiPicker/> component :", () => {
  test("display <EmojiPicker/> component ", async () => {
    let rendered = render(<EmojiPicker />);
    expect(rendered).toMatchSnapshot();
  });
});