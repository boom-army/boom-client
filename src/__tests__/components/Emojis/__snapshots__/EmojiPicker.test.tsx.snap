// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmojiPicker/> component : display <EmojiPicker/> component  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-s5gg9u"
      >
        <span
          class="emoji-pick"
        >
          <svg
            fill="#637584"
            height="18"
            viewBox="0 0 18 18"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 18C4.03702 18 0 13.963 0 9C0 4.03702 4.03702 0 9 0C13.963 0 18 4.03702 18 9C18 13.963 13.963 18 9 18ZM9 1.25581C4.73024 1.25581 1.25581 4.73024 1.25581 9C1.25581 13.2697 4.73024 16.7442 9 16.7442C13.2697 16.7442 16.7442 13.2697 16.7442 9C16.7442 4.73024 13.2697 1.25581 9 1.25581Z"
            />
            <path
              d="M8.99985 13.2824C7.41588 13.2824 5.95829 12.4871 5.10183 11.1525C4.91429 10.8611 4.99886 10.4744 5.29104 10.286C5.58239 10.0968 5.97086 10.1822 6.15839 10.4752C6.78379 11.4481 7.8462 12.0282 9.00075 12.0282C10.1552 12.0282 11.2177 11.4481 11.8439 10.4761C12.0314 10.1831 12.4199 10.0993 12.7112 10.2877C13.0043 10.4752 13.088 10.8637 12.9005 11.1551C12.0423 12.4895 10.5848 13.2857 9.00075 13.2857L8.99985 13.2824Z"
            />
            <path
              d="M11.2923 8.10923C11.9756 8.10923 12.5296 7.5552 12.5296 6.87183C12.5296 6.18843 11.9756 5.63443 11.2923 5.63443C10.6089 5.63443 10.0549 6.18843 10.0549 6.87183C10.0549 7.5552 10.6089 8.10923 11.2923 8.10923Z"
            />
            <path
              d="M6.70767 8.10923C7.39106 8.10923 7.94505 7.5552 7.94505 6.87183C7.94505 6.18843 7.39106 5.63443 6.70767 5.63443C6.02428 5.63443 5.47028 6.18843 5.47028 6.87183C5.47028 7.5552 6.02428 8.10923 6.70767 8.10923Z"
            />
          </svg>
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-s5gg9u"
    >
      <span
        class="emoji-pick"
      >
        <svg
          fill="#637584"
          height="18"
          viewBox="0 0 18 18"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 18C4.03702 18 0 13.963 0 9C0 4.03702 4.03702 0 9 0C13.963 0 18 4.03702 18 9C18 13.963 13.963 18 9 18ZM9 1.25581C4.73024 1.25581 1.25581 4.73024 1.25581 9C1.25581 13.2697 4.73024 16.7442 9 16.7442C13.2697 16.7442 16.7442 13.2697 16.7442 9C16.7442 4.73024 13.2697 1.25581 9 1.25581Z"
          />
          <path
            d="M8.99985 13.2824C7.41588 13.2824 5.95829 12.4871 5.10183 11.1525C4.91429 10.8611 4.99886 10.4744 5.29104 10.286C5.58239 10.0968 5.97086 10.1822 6.15839 10.4752C6.78379 11.4481 7.8462 12.0282 9.00075 12.0282C10.1552 12.0282 11.2177 11.4481 11.8439 10.4761C12.0314 10.1831 12.4199 10.0993 12.7112 10.2877C13.0043 10.4752 13.088 10.8637 12.9005 11.1551C12.0423 12.4895 10.5848 13.2857 9.00075 13.2857L8.99985 13.2824Z"
          />
          <path
            d="M11.2923 8.10923C11.9756 8.10923 12.5296 7.5552 12.5296 6.87183C12.5296 6.18843 11.9756 5.63443 11.2923 5.63443C10.6089 5.63443 10.0549 6.18843 10.0549 6.87183C10.0549 7.5552 10.6089 8.10923 11.2923 8.10923Z"
          />
          <path
            d="M6.70767 8.10923C7.39106 8.10923 7.94505 7.5552 7.94505 6.87183C7.94505 6.18843 7.39106 5.63443 6.70767 5.63443C6.02428 5.63443 5.47028 6.18843 5.47028 6.87183C5.47028 7.5552 6.02428 8.10923 6.70767 8.10923Z"
          />
        </svg>
      </span>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
